### PR TITLE
Add support for report generator plugins that process modified files as a batch

### LINKED ISSUE #299: Add support for report generator plugins that process modified files as a batch, rather than individually
SQLFluff implements the `diff-quality` plugin protocol. SQLFluff users have requested that the `diff-quality` integration should take advantage of SQLFluff's multiprocessing capability (processing files in parallel across multiple processes).

In order to do this, `BaseReportGenerator._diff_violations()` would need to be modified to call a different function if provided by the plugin, e.g. `violations_batch()` rather than `violations()`. If `violations_batch()` is not implemented, fall back to using `violations()`.

I'm happy to provide a PR to implement this, but wanted to ask if this is a reasonable feature and if there are any concerns with the proposed implementation. 🙏🏽

### PR DESCRIPTION
**Pull request by barrywhart**
Fixes #299

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -64,7 +64,8 @@ To install the development version:
 
     git clone https://github.com/Bachmann1234/diff-cover.git
     cd diff-cover
-    python setup.py install
+    poetry install
+    poetry shell
```
Comment (barrywhart) [inline README.rst:68]: Unrelated to the rest of the PR -- I noticed that the instructions mention `setup.py`, which no longer exists.

**Comment by Bachmann1234** [comment]
I'll look at this closely this week. Sorry I did not get to it yet

**Comment by barrywhart** [comment]
No problem! It's a roadmap item for us, not urgent. Hope you had a great holiday (if you're in the U.S.).

File: diff_cover/report_generator.py [hunk 2]
```diff
@@ -173,13 +173,22 @@ def _diff_violations(self):
         To make this efficient, we cache and reuse the result.
         """
         if not self._diff_violations_dict:
+            violations = None
+            src_paths_changed = self._diff.src_paths_changed()
+            try:
+                # Use violations_batch() if it's implemented.
```
Comment (Bachmann1234) [inline diff_cover/report_generator.py:179]: So this comment feels redundant to me

File: diff_cover/report_generator.py [hunk 3]
```diff
@@ -173,13 +173,22 @@ def _diff_violations(self):
...
+                violations = self._violations.violations_batch(src_paths_changed)
+            except NotImplementedError:
+                pass
             self._diff_violations_dict = {
                 src_path: DiffViolations(
-                    self._violations.violations(src_path),
+                    violations.get(src_path, [])
+                    if violations is not None
```
Comment (Bachmann1234) [inline diff_cover/report_generator.py:186]: so the violations reporter is supposed to cache these results so each is only processed once.

The function you are calling `violations_batch` could also be described as `warm_cache`.

Does that sound right to you? The reason I wanna rename this is because I want to hint that the implementor should be saving the results so we dont have to remember the state later (basically remove this if check)

basically....

Im wondering if you would go with renaming `violations_batch` to `warm_cache` and then removing the `if violations is not None` and just calling `violations` as is (under the assumption that the implementor has saved the results so the lookup is quick)

File: diff_cover/report_generator.py (hunk 2 above)
Comment (barrywhart) [inline diff_cover/report_generator.py:179]:
```suggestion
```

File: diff_cover/report_generator.py (hunk 3 above)
Comment (barrywhart) [reply diff_cover/report_generator.py:185]: My assumption in checking for ` if violations is not None` is that it will be `None` if the function is not implemented. In that case, wouldn't we need to call `violations()`?

Bigger picture: For SQLFluff, `violations_batch()` is not a warm cache. Rather, SQLFluff is a fairly slow-running tool that is often used to process hundreds or even thousands of files. If `diff-quality` calls SQLFluff one file at a time, there is no opportunity to use parallel processing. Adding a function like `violations_batch()` provides more opportunity for the implementor to use parallel processing or other clever strategies to improve performance. It can definitely be used for caching as well.

Does this change your thoughts on the above?

It may be helpful (in terms of code clarity) to **remove** the line `violations = None` and change the subsequent code to:
```
            try:
                violations = self._violations.violations_batch(src_paths_changed)
            except NotImplementedError:
                violations = {src_path: self._violations.violations(src_path)}
            self._diff_violations_dict = {
                src_path: DiffViolations(
                    violations.get(src_path, []),
                    ...
```

This ensures that `violations` is always defined and not `None`.

File: diff_cover/report_generator.py (hunk 2 above)
Comment (barrywhart) [reply diff_cover/report_generator.py]: Removed

File: diff_cover/report_generator.py (hunk 3 above)
Comment (barrywhart) [reply diff_cover/report_generator.py:185]: My proposed code above doesn't work "as is" because it uses `src_path` outside the loop. I'll try to do something similar. I was trying to avoid looping over the list of files twice, but maybe that's the best option.

File: diff_cover/report_generator.py [hunk 4]
```diff
@@ -172,15 +172,29 @@ def _diff_violations(self):
 
         To make this efficient, we cache and reuse the result.
         """
+        src_paths_changed = self._diff.src_paths_changed()
```
Comment (barrywhart) [inline diff_cover/report_generator.py:175]: I suggest reviewing this file with "Hide whitespace" enabled in GitHub.

File: diff_cover/report_generator.py [hunk 5]
```diff
@@ -172,15 +172,29 @@ def _diff_violations(self):
...
+                self._diff_violations_dict = {
+                    src_path: DiffViolations(
+                        violations.get(src_path, []),
+                        self._violations.measured_lines(src_path),
+                        self._diff.lines_changed(src_path),
+                    )
+                for src_path in src_paths_changed
+                }
+            except NotImplementedError:
+                self._diff_violations_dict = {
```
Comment (barrywhart) [inline diff_cover/report_generator.py:190]: This code is essentially identical to `main` (only difference is that we're looping over `src_paths_changed`.

File: diff_cover/report_generator.py [hunk 6]
```diff
@@ -172,15 +172,29 @@ def _diff_violations(self):
...
         """
+        src_paths_changed = self._diff.src_paths_changed()
         if not self._diff_violations_dict:
-            self._diff_violations_dict = {
-                src_path: DiffViolations(
-                    self._violations.violations(src_path),
-                    self._violations.measured_lines(src_path),
-                    self._diff.lines_changed(src_path),
+            try:
+                violations = self._violations.violations_batch(
```
Comment (barrywhart) [inline diff_cover/report_generator.py:178]: This code is very similar to the `except` block, but it uses results from `violations_batch()` rather than calling `violations()`.

It duplicates some code from the `except` block, but overall the duplication seems to make the code more readable by keeping the two implementations (using `violations_batch()` versus `violations()`) distinct from one another.

**Comment by barrywhart** [comment]
Ready for another review, I think.

File: pyproject.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "diff_cover"
-version = "7.1.0"
+version = "7.1.1"
```
Comment (Bachmann1234) [inline pyproject.toml:3]: let me handle this. I just do this when I bump versions. 

This is a feature and non breaking so I would probably go ahead and make this release 7.2.0 anyway

**Comment by Bachmann1234** [comment]
Ok. Think this can work. I think if you just update from main (and just remove the change to `pyproject.toml` ill do that when I cut the release) ill go ahead and merge this

**Comment by Bachmann1234** [comment]
wait, im an admin, what am I doing. Illl fix the conflict

**Comment by Bachmann1234** [comment]
@barrywhart I realized I should probably get a final "this is ready" from you before I merge and release. I tend to be a bit trigger happy with the release button. So just leave me a note and ill get this out asap (likely the evening I see it)

File: tests/test_report_generator.py [hunk 8]
```diff
@@ -50,10 +51,16 @@ class BaseReportGeneratorTest:
     SNIPPET_STYLE = ".css { color:red }"
     SNIPPET_TERMINAL = SNIPPET_MARKDOWN
 
+    # def violations_batch(*args, **kwargs):
+    #     raise NotImplementedError
+
```
Comment (barrywhart) [inline tests/test_report_generator.py:56]:
```suggestion
```

File: tests/test_report_generator.py [hunk 9]
```diff
@@ -50,10 +51,16 @@ class BaseReportGeneratorTest:
...
     @pytest.fixture(autouse=True)
     def base_setup(self, mocker):
         # Create mocks of the dependencies
-        self.coverage = mocker.MagicMock(BaseViolationReporter)
+        self.coverage = mocker.MagicMock(
+            BaseViolationReporter,
+            # violations_batch=self.violations_batch,
```
Comment (barrywhart) [inline tests/test_report_generator.py:62]:
```suggestion
```

**Comment by barrywhart** [comment]
I removed a few lines of commented-out code. This is ready to merge once the build passes.

Thanks for the review! SQLFluff users will appreciate the speedier performance once we update it to take advantage of this. 🙏

**Comment by Bachmann1234** [comment]
https://pypi.org/project/diff-cover/7.2.0/ is released!

**Comment by barrywhart** [comment]
That's great -- thank you so much!
